open-api/rest-catalog-open-api.yaml

```diff
@@ -3129,6 +3204,11 @@ components:
          - `s3.secret-access-key`: secret for credentials that provide access to data in S3 
          - `s3.session-token`: if present, this value should be used for as the session token 
          - `s3.remote-signing-enabled`: if `true` remote signing should be performed as described in the `s3-signer-open-api.yaml` specification
+
+        ## Credentials
+
+        Credentials for ADLS / GCS / S3 are provided through the `credentials` field. Clients should first check whether the
+        respective credentials exist in the `credentials` field before checking the `config` for credentials.
```

Comment:
A table may have metadata and data files spread out across multiple locations (because ultimately clients can append data from from any location). It would be nice to support those use cases too.

Reply:
A valid point. However, that could not just span e.g. 2 S3 buckets in the same object store, but across different object stores - even two S3 - one on Minio and one on Ceph - or GCS and S3 or....

Reply:
It is tricky that the arbitrary file locations can easily break the boundary of credentials.

Reply:
Not too tricky, I hope :) WDYT about mapping base table URIs to credentials? The server could return this mapping in response to the client's request for credentials, then the client would choose the appropriate credential for each file it needs to access.

If the client provided some more information about the time of operation it needs to perform (e.g. read the latest snapshot or append new files to the latest snapshot) the server should be able to return a fairly narrow set (1 in most cases) of credentials.

Reply:
> A table may have metadata and data files spread out across multiple locations (because ultimately clients can append data from from any location). It would be nice to support those use cases too.

I don't think this should be the scope of this PR and rather warrants a separate proposal

Reply:
If a more feature-rich credentials endpoint spec is in the works (as discussed in another comment), I agree, it would be preferable to move this discussion there. However, I think it makes sense to tackle this problem during together with the dedicated credentials endpoint.

Reply:
Credential policies can be created to accommodate multiple locations/buckets, so that is a server-side issue, not related to the credentials.  If I have multiple buckets, the policy just needs to be crafted such that appropriate access is granted for the specified locations.  There's no need for a mapping.

Reply:
@danielcweeks : does your comment assume all locations are within with same storage provider?

Reply:
That does currently and I don't think multiple would be supported by the implementation as of today.  We could possibly allow multiple credentials (one per provider?), but I think we should require that the provided credential be scoped for all table locations within a provider and not push that mapping onto the client since all of the policies I'm aware of allow for that.

@nastra I think multiple provider is an interesting question, but also complicates this.

Reply:
I did not mean to overload this PR with tables split across diverse locations, but it would be nice to consider that use case in the follow-up spec revision with a dedicated credentials endpoint (as @nastra mentioned in another comment on this PR).

Reply:
I mean technically we could design the spec/impl in a way that would allow a server to sent back credentials for multiple providers (one credential per provider) in a single response. So rather than having
```
"credentials" : {
    "type" : "gcs",
    "token" : "gcsToken",
    "expires-at-ms" : 1000
  }
```

we could have 
```
"credentials" : {
  "s3": {
    "access-key-id" : "keyId",
    "secret-access-key" : "accessKey",
    "session-token" : "sessionToken",
    "expires-at-ms" : 1000
  },
  "gcs": {
    "token" : "gcsToken",
    "expires-at-ms" : 1000
  },
  // no adls credentials as that provider isn't used for that table
} 
```

Also I agree that we most likely don't want to push the mapping to the client. Is that what you guys had in mind?

Reply:
> we most likely don't want to push the mapping to the client

How can the client deal with multiple credentials without any mapping? Even sending one credential per provider already implies mapping by provider type... Does it not?

A credential per provider would be a good start.

Reply:
we could make `credentials` return a list of any type of credential and each credential could have a `scheme` or `prefix` property, which can then be used to map that credential, wdyt?

Reply:
That sounds reasonable to me :)

nit: I'd rename `scheme` to `base` or something like that... and maybe make it optional, because `scheme == gs` is essentially the same as `type == gcs`, I guess.

Reply:
it's probably ok to rename `scheme` to `prefix` here and make it optional